src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp

```diff
@@ -0,0 +1,237 @@
+    {
+        _failures++;
+        printf("FAIL: ICorProfilerInfo::GetEnvironmentVariable() failed hr=0x%x", hr);
+        return hr;
+    }
+    String envVarStr = envVar;
+    String expectedEnvVarStr = reinterpret_cast<const WCHAR*>(u"TRUE");
+    if (envVarStr == expectedEnvVarStr)
+    {
+        printf("Setting GarbageCollectionStarted event masks\n");
+        eventMask |= COR_PRF_MONITOR_GC;
+    }
+
+    if (FAILED(hr = pCorProfilerInfo->SetEventMask2(eventMask, COR_PRF_HIGH_MONITOR_NONE)))
+    {
+        _failures++;
+        printf("FAIL: ICorProfilerInfo::SetEventMask2() failed hr=0x%x", hr);
+    }
+
+    return hr;
+}
+
+HRESULT STDMETHODCALLTYPE GCHeapEnumerationProfiler::GarbageCollectionStarted(int cGenerations, BOOL generationCollected[], COR_PRF_GC_REASON reason)
+{
+    SHUTDOWNGUARD();
+    printf("GCHeapEnumerationProfiler::GarbageCollectionStarted\nSleeping for 10 seconds\n");
+    std::this_thread::sleep_for(std::chrono::seconds(10));
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE GCHeapEnumerationProfiler::GarbageCollectionFinished()
+{
+    SHUTDOWNGUARD();
+    printf("GCHeapEnumerationProfiler::GarbageCollectionFinished\n");
+    return S_OK;
+}
+
+HRESULT GCHeapEnumerationProfiler::Shutdown()
+{
+    Profiler::Shutdown();
+
+    if (_expectedExceptions == 1)
+    {
+        printf("GCHeapEnumerationProfiler::Shutdown: PASS: Encountered exception as expected.\n");
+        printf("PROFILER TEST PASSES\n");
+        fflush(stdout);
+        return S_OK;
+    }
+
+    if (_objectsCount < 100)
+    {
+        printf("GCHeapEnumerationProfiler::Shutdown: FAIL: Expected at least 100 objects, got %d\n", _objectsCount.load());
+        _failures++;
+    }
+
+    if (_customGCHeapObjectTypesCount != 1)
+    {
+        printf("GCHeapEnumerationProfiler::Shutdown: FAIL: Expected 1 custom GCHeapObject type, got %d\n", _customGCHeapObjectTypesCount.load());
+        _failures++;
+    }
+
+    if (_failures == 0)
+    {
+        printf("PROFILER TEST PASSES\n");
+    }
+    else
+    {
+        // failures were printed earlier when _failures was incremented
+    }
+    fflush(stdout);
+
+    return S_OK;
+}
+
+String GCHeapEnumerationProfiler::GetClassIDNameHelper(ClassID classId) {
+    return GetClassIDName(classId);
+}
+
+struct CallbackState
+{
+    std::atomic<int>* objectsCount;
+    GCHeapEnumerationProfiler *instance;
+    std::atomic<int>* customGcHeapObjectTypeCount;
+};
+
+static BOOL STDMETHODCALLTYPE heap_walk_fn(ObjectID object, void* callbackState)
+{
+    CallbackState* state = static_cast<CallbackState*>(callbackState);
+
+    state->objectsCount->fetch_add(1, std::memory_order_relaxed);
+
+    ClassID classId{0};
+    HRESULT hr = state->instance->pCorProfilerInfo->GetClassFromObject(object, &classId);
+    if (hr != S_OK)
+    {
+        printf("Error: failed to get class ID from object.\n");
```

Comment:
I think we should expect that every object has an obtainable class, if not the test should fail. This prints out a console error message but it doesn't appear to record anything that makes the test fail.

Reply:
Made the failure counter publicly incrementable, and changed the walk function to stop enumerating the GC Heap if a class could not be resolved.